### ISSUE TITLE
[next-js] setImmediate is not defined

### ISSUE THREAD

**Issue by nandorojo** [expo/expo#7996]
### Environment

```sh
  Expo CLI 3.18.6 environment info:
    System:
      OS: macOS 10.15.2
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 12.13.1 - /usr/local/bin/node
      Yarn: 1.22.4 - /usr/local/bin/yarn
      npm: 6.12.1 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Xcode: 11.3/11C29 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~37.0.3 => 37.0.7 
      react: ~16.9.0 => 16.9.0 
      react-native: https://github.com/expo/react-native/archive/sdk-37.0.1.tar.gz => 0.61.4 
    npmGlobalPackages:
      expo-cli: 3.18.6
```

expo + next.js app (SDK 37)

### Steps to Reproduce

Create a next.js project with expo, import `react-native-reanimated`, use it in a component, and you'll see **Error was not caught ReferenceError: setImmediate is not defined.** in the browser console errors.

This also might be related to `expo-web-browser` being imported. I haven't dug too deep, but I found the solution and will put it at the bottom of this issue.

### Expected Behavior

App works.

### Actual Behavior

App throws error:

<img width="545" alt="Screen Shot 2020-04-24 at 3 26 22 PM" src="https://user-images.githubusercontent.com/13172299/80249655-3a28c200-8640-11ea-9dc3-1057d18b22ee.png">

```sh
Error was not caught ReferenceError: setImmediate is not defined
```

**Comment by nandorojo** [comment]
# Solution

Polyfill `setImmediate`, which is used by `react-native-reanimated`.

```sh
yarn add setimmediate
```

In `pages/_app.tsx`, at the top of the file:

```js
import 'setimmediate'
```

Here's my `pages/_app.tsx` file, for reference:

```js
import React from 'react'
import { AppProps } from 'next/app'
import Providers from '../src/providers'
import styled from 'styled-components/native'
import { ThemeProps } from '../src/theme'
import Header from '../src/components/Header.web'

// fix the setImmediate usage from react-native-reanimated
import 'setimmediate'

if (process.browser) {
  if (location.protocol !== 'https:' && process.env.NODE_ENV !== 'production') {
    console.warn(
      'This app requires HTTPS. Try cancelling this session and running yarn https.'
    )
  } else {
    const WebBrowser = require('expo-web-browser')
    WebBrowser.maybeCompleteAuthSession()
  }
}

const Wrapper = styled.View`
  flex: 1;
  background-color: ${({ theme }: ThemeProps) => theme.colors.background};
`

export default ({ Component, pageProps, router }: AppProps) => {
  return (
    <Providers>
      <Header />
      <Wrapper>
        <Component {...pageProps} key={router.route} />
      </Wrapper>
    </Providers>
  )
}
```

I figured I'd post this for anyone who finds themselves here. I bet I'll also run into this again someday, so this is for future me as well.

**Comment by emroot** [comment]
that's interesting. thanks for sharing. It worked for me, but only in development. With a production build `next build`, it breaks on safari and iOS browsers. When I first load the page it works fine, however if I refresh the page, I get `setimmediate is not defined`. Investigating what's up.

**Comment by nandorojo** [comment]
Try this:

```js
if (!global.setImmediate) {
  global.setImmediate = setTimeout
}
```

**Comment by emroot** [comment]
I'm not sure why, but adding the polyfill in a different place worked. 
I'm importing the polyfill right before I make my call to `react-native-reanimated`, lower down the chain instead of `pages/_app.tsx` and it fixes the issue.

**Comment by nandorojo** [comment]
Oh, got it. I did this in my `pages/_app.tsx` before importing reanimated, since the pollyfill didn’t always work for me.

**Comment by nandorojo** [comment]
To this day I still get this bug when the site builds and publishes, but never in build mode. It feels completely random, since it only happens on some deploys. The best work-around I've found is to lazy load some components with `next/dynamic` that use reanimated, and to do this at the top of pages:

```js
import 'setimmediate'
if (!global.setImmediate) {
  global.setImmediate = setTimeout
}
```

**Comment by ezekiel747** [comment]
Hi, i'm having a somewhat similar issue, but with requestAnimationFrame instead of setImmediate.
Using latest Expo + Nextjs + Drawer navigator from react-navigation (v5)
For dev mode, the workaround is fine (setting global.setRequestAnimationFrame to setTimeout for example).
But when trying run "next build" i get:

_info  - Creating an optimized production build  
info  - Compiled successfully

> Build error occurred
ReferenceError: requestAnimationFrame is not defined_

Tried several ways, i can't seem to get past this issue.
I would appreciate any help. Thank you in advance!

**Comment by yuta-hidaka** [comment]
@ezekiel747 

Hi I got same issue.

In my case I install `raf` and put this code top of the `/pages/index,tsx`(or index.js or _app.js)

install modules `yarn add raf` or `npm install raf`

```
import 'setimmediate'; // put on top of file
import 'raf/polyfill' // put on top of file

export default function App() {
  return (
      <YOUR COMPONENT>
  );
}

```

**Comment by nandorojo** [comment]
All you need to do now is this in `pages/_app.js`:

```js
import 'raf/polyfill'
```

Also be sure to upgrade to Reanimated 2.1.0, which no longer uses setImmediate on web. You'll need to do one more patch when you get `window is not defined`, which I made a PR for on Reanimated's repo already.

**Comment by nandorojo** [comment]
https://github.com/software-mansion/react-native-reanimated/pull/1891/files

I would just use that code with `patch-package` until the next release.

**Comment by ezekiel747** [comment]
I'm trying to use patch-package but I'm a bit confused.
I've replicated the changes from the PR (in the src folder), generated the patch using patch-package,
but i'm still getting the same error, which seems to be in the corresponding js file from the lib folder, instead of src.

Do i need to reinstall the package? It isn't a direct dependency, it's used only by @react-navigation/drawer.
I tried removing the react-native-reanimation folder, then reinstall the drawer package. Same issue.

I guess i need to patch the transpiled js files in the lib instead?
Thanks in advance.
